### ISSUE TITLE
set_proc_address from the README doesn't compile

### ISSUE THREAD

**Issue by MagpieMcGraw** [vassvik/odin-gl#2]
I don't yet know function pointers in Odin, so I just copied the code snipped from the readme. However, it doesn't compile.
![screenshot at 2018-02-02 19-12-01](https://user-images.githubusercontent.com/17641521/35745847-59000cf6-084d-11e8-895f-cc00c3cc443c.png)

I'm not really sure what's supposed to happen here. I assume that glfw.GetProcAddress retrieves a pointer to a function. And then it get's assigned to the pointer(p) that was passed to set_proc_address. But it seems you have to do some kind of complicated cast for it to work. Which doesn't compile. And I tried removing all the casts, at which point things compile, but segfault.

**Comment by vassvik** [comment]
Thank you for your notice. I have updated the README to reflect the updated version which uses slightly newer syntax. 

In summary, you have three options on how to fetch a function pointer from the driver: 
```go
// assigning directly, but casting the result of glfw.GetProcAddress
CullFace = cast(proc"c"(u32))glfw.GetProcAddress("glCullFace");

// passing a pointer to the procedure variable to a set_proc_address procedure, but interpreting it as a rawptr.
set_proc_address :: proc(p: rawptr, string: name) {
	(cast(^rawptr)p)^ = glfw.GetProcAddress(&name[0]);
}
set_proc_address(&CullFace, "glCullFace\x00");

// passing a pointer to the procedure variable to a set_proc_address procedure, but interpreting it as a ^rawptr
set_proc_address :: proc(p: ^rawptr, string: name) {
	p^ = glfw.GetProcAddress(&name[0]);
}
set_proc_address(cast(^rawptr)&CullFace, "glCullFace\x00");
```
Arguably the second option looks better when you have to fetch function pointers in bulk.


More lengthy explanation below, which I will make into a wiki post on the Odin repo fairly soon. 

---

Modern OpenGL functions has to be loaded at runtime from the dynamic/shared library supplied by your driver (`opengl32.lib` in Windows and `libGL.so` in Linux). To use them we need to declare *all* OpenGL functions we want to use as function pointers, then we call the appropriate OS API to fetch the function pointers from driver at runtime. The last part is typically done by calling a function like `wglGetProcAddress` (in Windows), or `glXGetProcAddress` (in Linux), or an OS wrapper function like `glfwGetProcAddress` with the function name as a string for input. 

Consider the first function listed in `gl.odin`, `glCullFace`. It is declared as: 
```go
CullFace: proc "c" (mode: u32);
```
and is simply just a procedure variable (equivalent to a function pointer in C). 

Say I wanted to use `glfw.GetProcAddress` to grab the function pointers from the driver. I could then assign to the `CullFace` variable like 
```go
CullFace = cast(proc"c"(u32))glfw.GetProcAddress("glCullFace");
```

But this is a bit unwieldy to handle for a huge library such as gl.odin (a lot of extra typing for the explicit casting). Instead we can pass a pointer to the procedure variable as a parameter to a `set_proc_address` procedure that sets it for us
```go
set_proc_address :: proc(p: rawptr, string: name) {
	(cast(^rawptr)p)^ = glfw.GetProcAddress(&name[0]);
}
set_proc_address(&CullFace, "glCullFace\x00");
```
where the casting of `p` to a pointer reflects that the input to the `set_proc_address` is a *pointer to a function pointer*!

The reason one makes the `p` parameter a `rawptr` instead of a pointer to a `rawptr` is that in that case one would have to cast upon calling `set_proc_address`, which looks worse. 
```go
set_proc_address :: proc(p: ^rawptr, string: name) {
	p^ = glfw.GetProcAddress(&name[0]);
}
set_proc_address(cast(^rawptr)&CullFace, "glCullFace\x00");
```
Every pointer type can be assigned to a `rawptr` variable, but not a `^rawptr` (which is *not* a `rawptr`!).

**Comment by MagpieMcGraw** [comment]
I looked in Bill's source code and found a set_proc_address that compiles, but my program kept segfaulting. After some more looking, I also found a set_proc_address in glfw.odin, but the program kept segfaulting anyway. Eventually I figured out that I was loading functions from gl.odin, but trying to call functions from opengl.odin. Ooops.

About this line: (cast(^rawptr)p)^ = glfw.GetProcAddress(&name[0]);
We cast p to a (pointer to a rawptr). And the we use ^ to travel to the rawptr pointed to, and write the result of GetProcAddress. Right?

Anyway, everything appears to work now. Thanks!
